pkg/dogstatsd/mapper/mapper.go

```diff
@@ -0,0 +1,124 @@
+
+package mapper
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+)
+
+var (
+	allowedGlobMatchPattern = regexp.MustCompile(`^[a-zA-Z0-9\-_*.]+$`)
+)
+
+const (
+	matchTypeGlob  = "glob"
+	matchTypeRegex = "regex"
+)
+
+// MetricMapper contains mappings and cache instance
+type MetricMapper struct {
+	Mappings []MetricMapping
+	cache    *mapperCache
+}
+
+// MetricMapping represent one mapping rule
+type MetricMapping struct {
+	Match     string            `mapstructure:"match"`
+	MatchType string            `mapstructure:"match_type"`
+	Name      string            `mapstructure:"name"`
+	Tags      map[string]string `mapstructure:"tags"`
+	regex     *regexp.Regexp
+}
+
+// NewMetricMapper creates, validates, prepares a new MetricMapper
+func NewMetricMapper(configMappings []MetricMapping, cacheSize int) (MetricMapper, error) {
+	var mappings []MetricMapping
+	for i := range configMappings {
+		currentMapping := configMappings[i]
+		if currentMapping.MatchType == "" {
+			currentMapping.MatchType = matchTypeGlob
+		}
+		if currentMapping.MatchType != matchTypeGlob && currentMapping.MatchType != matchTypeRegex {
+			return MetricMapper{}, fmt.Errorf("mapping num %d: invalid match type, must be `glob` or `regex`", i)
+		}
+		if currentMapping.Name == "" {
+			return MetricMapper{}, fmt.Errorf("mapping num %d: name is required", i)
+		}
+		if currentMapping.Match == "" {
+			return MetricMapper{}, fmt.Errorf("mapping num %d: match is required", i)
+		}
+		err := currentMapping.prepare()
+		if err != nil {
+			return MetricMapper{}, err
+		}
+		mappings = append(mappings, currentMapping)
+	}
+	cache, err := newMapperCache(cacheSize)
+	if err != nil {
+		return MetricMapper{}, err
+	}
+	return MetricMapper{Mappings: mappings, cache: cache}, nil
+}
+
+// prepare compiles the match patterns into regexes
+func (m *MetricMapping) prepare() error {
+	metricRe := m.Match
+	if m.MatchType == matchTypeGlob {
+		if !allowedGlobMatchPattern.MatchString(m.Match) {
+			return fmt.Errorf("invalid glob match pattern `%s`, it does not match allowed match regex `%s`", m.Match, allowedGlobMatchPattern)
+		}
+		if strings.Contains(m.Match, "**") {
+			return fmt.Errorf("invalid glob match pattern `%s`, it should not contain consecutive `*`", m.Match)
+		}
+		metricRe = strings.Replace(metricRe, ".", "\\.", -1)
+		metricRe = strings.Replace(metricRe, "*", "([^.]*)", -1)
+	}
+	regex, err := regexp.Compile("^" + metricRe + "$")
+	if err != nil {
+		return fmt.Errorf("invalid match `%s`. cannot compile regex: %v", m.Match, err)
+	}
+	m.regex = regex
+	return nil
+}
+
+// GetMapping returns:
+// - name: the mapped expanded name
+// - tags: the tags extracted from the metric name and expanded
+// - matched: weather we found a match or not
+func (m *MetricMapper) GetMapping(metricName string) (string, []string, bool) {
+	result, cached := m.cache.get(metricName)
+	if cached {
+		return result.Name, result.Tags, result.Matched
+	}
+	for _, mapping := range m.Mappings {
```

Comment:
I couldn't be on the RFC you've presented but design question regarding performance here.
This whole logic is only applied if some mappings are defined in the config and only on metrics that are submitted without tags. This is great, couple questions though

1. Adding a mapper for just a few metrics (like airflow) will impact performance for any other statsd-type metric even though the dogstatsd service might already be heavily used. What do you think of branching out of this logic before? I expect that we'll have many mapper for each integration, could we imagine having a first matcher like "does my metric name contains 'airflow'" and if so apply the airflow mapper. Just an rough idea but having a way to limit the performance impact for the metric that don't require mappers.
2. You've run some benchmarks before but if my understanding is correct, the more mapping we add the more we expect performance to get worse (and in a linear way). That means that if 10 integration require 10 mappers each, we will check every single metric against all 100 mappers. Same rough idea as before, can we branch out before so that we just run the regex that are relevant?

All of that depends on one difficult aspect though, is it possible to differentiate metric that will require a mapper from those that don't without applying each pattern? There might be some easy things like having 'airflow' as the prefix, and can we rely on this in a general way.


Maybe something like this in the configuration and allowing `*` mappers that will be applied to every single metric?
```yaml
dogstatsd_mappings:
  'airflow':
    - match: <METRIC_TO_MATCH>
       match_type: glob
       name: <METRIC_NAME>
       ...
  'vault':
   - match: <METRIC_TO_MATCH>
      ...
  '*':
    - match: <METRIC_TO_MATCH>
       ....
```

Reply:
IIUC, your idea is to add a step to match the prefix before the actual matching.

Since we are using a LRU cache, each metric is only parsed once and then cached and the cache have a complexity of O(1). So, the performance should not decrease in the visible way since the mapping logic starting from `for _, mapping := range m.Mappings` is only executed once per metric.

Matching the prefix beforehand won't reduce much the performance but should significantly reduce the number of items cached if used before the cache retrieval (but might reduce a bit the perf since we need to check for prefix, then check in the cache).

Keeping the number of items in the cache low could be crucial since the cache key is a `statsd` metric name. The number of incoming entries can be high since `statsd` metric can contain elements like `job_name`, `task_id`, etc

So overall, I think it's a good thing to match the prefix. A little twist to your proposition. I would add a concept of `mapping_group` or maybe `profile` (like snmp), this will add flexibility in case we need other configuration later beside `prefix` at group level.

```
dogstatsd_mappings:
  # mapping_group 1
  - prefix: 'airflow.'
    mappings: 
      - match: <METRIC_TO_MATCH>
         match_type: glob
         name: <METRIC_NAME>
         ...
  # mapping_group 2
  - prefix: 'vault.'
    mappings: 
      - match: <METRIC_TO_MATCH>
         ...
      ...
  # mapping_group 3
  # match all if no prefix
  - mappings: 
      - match: <METRIC_TO_MATCH>
         ...
      ...

```